src/stub/parser.rs

```diff
@@ -45,6 +45,27 @@ impl<'a> Parser<'a> {
             };
         }
 
+        let mut read_pairings = Vec::new();
+        for read_cmd in &stub.commands {
+            if let Cmd::Read(var_cmds) = read_cmd {
+                for var_cmd in var_cmds {
+                    read_pairings.push((var_cmd.ident.clone(), var_cmd.var_type.clone()));
+                }
+            }
+        }        
+
+        for cmd in &mut stub.commands {
+            if let Cmd::WriteJoin { join_terms, output_comment: _ } = cmd {
+                for term in join_terms.iter_mut() {
+                    for (ident, var_type) in &read_pairings {
+                        if term.ident == *ident {
+                            term.var_type = Some(var_type.clone());
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
This probably makes no difference but at least we can get rid of one of the loops.
I also added a `#[derive(Copy)]` for the `VarType` enum so there's no need to explicitly .clone() `var_type` (copying enums without complicated data is cheap).

```suggestion
        let mut read_pairings = std::collections::BTreeMap::new();
        for read_cmd in &stub.commands {
            if let Cmd::Read(var_cmds) = read_cmd {
                for var_cmd in var_cmds {
                    read_pairings.insert(var_cmd.ident.clone(), var_cmd.var_type);
                }
            }
        }        

        for cmd in &mut stub.commands {
            if let Cmd::WriteJoin { join_terms, output_comment: _ } = cmd {
                for term in join_terms.iter_mut() {
                    if let Some(var_type) = read_pairings.get(&term.ident) {
                        term.var_type = Some(*var_type);
                    }
                }
            }
        }
```

Reply:
Sure, I added it in [57ef078](https://github.com/ellnix/clash/pull/30/commits/57ef0788d99357b5e97b572e687d1723d7d5020e)